File: src/main/java/org/gjt/jclasslib/structures/elementvalues/ConstElementValue.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/
package org.gjt.jclasslib.structures.elementvalues;

import org.gjt.jclasslib.structures.InvalidByteCodeException;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

/**
 * Describes an  <tt>ConstElementValue</tt> attribute structure.
 *
 * @author <a href="mailto:vitor.carreira@gmail.com">Vitor Carreira</a>
 *
 */
public class ConstElementValue extends ElementValue {

    public final static String ENTRY_NAME = "ConstElement";

    private static final int LENGTH = 2;
    private int constValueIndex;

    protected ConstElementValue(int tag) {
        super(tag);
    }

    /**
     * Get the <tt>const_value_index</tt> of this element value entry.
     *
     * @return the <tt>const_value_index</tt>
     */
    public int getConstValueIndex() {
        return this.constValueIndex;
    }

    /**
     * Set the <tt>const_value_index</tt> of this element value entry.
     *
     * @param constValueIndex the <tt>const_value_index</tt>
     */
    public void setConstValueIndex(int constValueIndex) {
        this.constValueIndex = constValueIndex;
    }

    protected int getSpecificLength() {
        return LENGTH;
    }

    public void read(DataInput in) throws InvalidByteCodeException, IOException {
        super.read(in);

        constValueIndex = in.readUnsignedShort();

        if (debug) debug("read ");
    }

    public void write(DataOutput out) throws InvalidByteCodeException, IOException {
        super.write(out);

        out.writeShort(constValueIndex);

        if (debug) debug("wrote ");
    }

    protected void debug(String message) {
        super.debug(message +
                "ConstElementValue with const_value_index " +
                constValueIndex);
    }

    public String getEntryName() {
        return ENTRY_NAME;
    }
}

File: modules/data/src/main/kotlin/org/gjt/jclasslib/structures/elementvalues/ConstElementValue.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/
package org.gjt.jclasslib.structures.elementvalues

import java.io.DataInput
import java.io.DataOutput

/**
 * Describes a ConstElementValue attribute structure.
 */
class ConstElementValue(elementValueType: ElementValueType) : ElementValue(elementValueType) {
    /**
     * const_value_index of this element value entry.
     */
    var constValueIndex: Int = 0

    override val length: Int
        get() = super.length + 2

    override fun readData(input: DataInput) {
        constValueIndex = input.readUnsignedShort()
    }

    override fun writeData(output: DataOutput) {
        super.writeData(output)
        output.writeShort(constValueIndex)
    }

    override val debugInfo: String
        get() = "with constValueIndex $constValueIndex"

    override val entryName: String
        get() = "ConstElement"

}
